pulser/simulation/simulation.py

```diff
@@ -15,116 +15,302 @@
-        """Initialize the Simulation with a specific pulser.Sequence."""
-        supported_bases = {"ground-rydberg", "digital"}
+                 config: Optional[SimConfig] = None,
+                 evaluation_times: Union[float, str, ArrayLike] = 'Full'
+                 ) -> None:
+        """Instantiates a Simulation object."""
         if not isinstance(sequence, Sequence):
             raise TypeError("The provided sequence has to be a valid "
                             "pulser.Sequence instance.")
         if not sequence._schedule:
             raise ValueError("The provided sequence has no declared channels.")
+        if all(sequence._schedule[x][-1].tf == 0 for x in sequence._channels):
+            raise ValueError("No instructions given for the channels in the "
+                             "sequence.")
         not_supported = (set(ch.basis for ch in sequence._channels.values())
-                         - supported_bases)
+                         - SUPPORTED_BASES)
         if not_supported:
             raise NotImplementedError("Sequence with unsupported bases: "
                                       + "".join(not_supported))
-        if all(sequence._schedule[x][-1].tf == 0 for x in sequence._channels):
-            raise ValueError("No instructions given for the channels in the "
-                             "sequence.")
         self._seq = sequence
         self._qdict = self._seq.qubit_info
         self._size = len(self._qdict)
         self._tot_duration = max(
             [self._seq._last(ch).tf for ch in self._seq._schedule]
         )
-
         if not (0 < sampling_rate <= 1.0):
             raise ValueError("The sampling rate (`sampling_rate` = "
                              f"{sampling_rate}) must be greater than 0 and "
                              "less than or equal to 1.")
-        if int(self._tot_duration*sampling_rate) < 4:
+        if int(self._tot_duration * sampling_rate) < 4:
             raise ValueError("`sampling_rate` is too small, less than 4 data "
                              "points.")
-        self.sampling_rate = sampling_rate
-
+        self._sampling_rate = sampling_rate
         self._qid_index = {qid: i for i, qid in enumerate(self._qdict)}
         self.samples: dict[str, dict[str, dict]] = {
             addr: {basis: {} for basis in ['ground-rydberg', 'digital']}
-            for addr in ['Global', 'Local']
-        }
+            for addr in ['Global', 'Local']}
         self.operators = deepcopy(self.samples)
-
+        self._collapse_ops: list[qutip.Qobj] = []
+        self._times = self._adapt_to_sampling_rate(
+            np.arange(self._tot_duration, dtype=np.double)/1000)
+        self.evaluation_times = evaluation_times
+        self._config = config if config else SimConfig()
+        self._update_noise()
         self._extract_samples()
         self._build_basis_and_op_matrices()
         self._construct_hamiltonian()
```

Comment:
Why not call `self._build_basis_and_op_matrices()` and then `self._update_hamiltonian` instead? It seems to be an encapsulation of these calls. 

Even better, considering that `self._construct_hamiltonian` is only called here and inside `self._update_hamiltonian`, what you can do is move the `self._update_noise()`, the reset of `self.samples` and `self._extract_samples()` to the top of `_construct_hamiltonian()` and use it instead of `_update_hamiltonian` throughout. The two names are a bit ambiguous anyway. So, this call would rather be:
```suggestion
        self._build_basis_and_op_matrices()
        self._construct_hamiltonian()
```
and then you replace all calls to `_update_hamiltonian` with `_construct_hamiltonian`.

Is this clear?

Reply:
The only issue is you need to `_extract_samples`once before running `_build_basis_and_op_matrices`since you need to know the type of basis needed, and that's given through the samples.